### PR TITLE
TTAHUB-27 Draft of proposed schema

### PR DESCRIPTION
**Pull request by gopar**
## Description of change

This draft is to propose changes to facilitate multiple approvers.

Here's my proposed solution.

For this ticket:
- Backend
  - Update the schema as shown in this PR. Data is **not** moved or transformed.
  - Add application logic that does the following:
    - endpoint "submit an activity report for approval" creates new row in `activityReportApprover` table. `activityReport.status` is still updated to be "submitted".
    - endpoint "review an activity report" now updates to `activityReportApprover.status` and `activityReportApprover.note` instead of updating cols in `activityReport`. `activityReport.status` stays "submitted".
    - endpoint "get activity report" requests will return the activity report, and a new field not stored in the database called `calculatedApprovalStatus`. `calculatedApprovalStatus` is determined using business logic stored in the application, it is not stored in the database. Possible values of `calculatedApprovalStatus` will be `null` (report has not been submitted for approval, report is in draft mode), "submitted" (no approver has taken action), "needs action" (any approver said it needs changes), and "approved" (all approvers approved). activity report will include an array of `activityReportApprover` objects. 
    - After review PR is held.
- Frontend is updated to expect new api responses, behind feature flag.
- Deployment with all BE changes and feature flag turned on.

Optional future ticket:
Create tool to query production db to select all AR with `activityReport.legacyApprovingManagerId`, export, and import data into `activityReportApprover` table. Set legacy data values to `null`. Set any `activityReport.status` with value "approved" to value "submitted". Create db migration to remove legacy fields.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-27

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: docs/logical_data_model.md [hunk 1]
```diff
@@ -154,7 +154,7 @@ class ActivityReport {
   startDate : date
   activityRecipientType : string
   requester : string
-  * status : string
+  * overviewStatus : string
```
Comment (gopar) [inline docs/logical_data_model.md:157]: This can be removed and computed in the client side if we choose to remove it.

Comment (dcloud) [reply docs/logical_data_model.md:157]: I agree. From looking at the issue, I don't think we'd need a field for this. I think the API could provide the necessary information as if it were part of the AR model itself.

**Review by dcloud — COMMENTED** [review]
I think this schema works, though I wonder if we'd want to do more to capture the actions/history related to status changes. One thing I noticed on [the ticket](https://ocio-jira.acf.hhs.gov/browse/TTAHUB-27) was discussion about being able to add/remove approvers after a report is submitted. The way we do things now we wouldn't capture that "approver X was added on X date and removed on Y date". That said, we don't record much metadata about various changes made to reports, so perhaps that's not something we need to consider here.

File: docs/logical_data_model.md (hunk 1 above)
Comment (SarahJaine) [reply docs/logical_data_model.md]: status will remain as is for legacy reasons, data can be transformed in a new ticket after multiple approvers is implemented

File: docs/logical_data_model.md [hunk 2]
```diff
@@ -163,7 +163,8 @@ class ActivityReport {
   ttaType : array<string>
   context : string
   pageState : json
-  managerNotes : string
+  legacyManagerNotes : string
+  legacyApprovingManagerId : integer(32) REFERENCES public.Users.id
```
Comment (SarahJaine) [inline docs/logical_data_model.md:167]: `approvingManagerId` should have been in this schema already. I'm proposing renaming this field `legacyApprovingManagerId`.

File: docs/logical_data_model.md [hunk 3]
```diff
@@ -163,7 +163,8 @@ class ActivityReport {
   ttaType : array<string>
   context : string
   pageState : json
-  managerNotes : string
+  legacyManagerNotes : string
```
Comment (SarahJaine) [inline docs/logical_data_model.md:166]: proposing renaming `managerNotes` to `legacyManagerNotes`

File: docs/logical_data_model.md [hunk 4]
```diff
@@ -154,7 +154,7 @@ class ActivityReport {
   startDate : date
   activityRecipientType : string
   requester : string
-  * status : string
+  * status : enum
```
Comment (SarahJaine) [inline docs/logical_data_model.md:157]: This was changed to an enum awhile back but the schema wasn't updated.

File: docs/logical_data_model.md (hunk 3 above)
Comment (dcloud) [reply docs/logical_data_model.md:166]: I think `legacy` could be confusing, given that label was given to reports imported from Smartsheet. If these fields are going away in the future after a data migration, then this isn't much of a concern.

File: docs/logical_data_model.md [hunk 5]
```diff
@@ -196,6 +197,14 @@ class ActivityReportCollaborator {
   * userId : integer(32) REFERENCES public.User.id
 }
 
+class ActivityReportApprover {
+  * id : integer <<generated>>
+  * activityReportId : integer(32) REFERENCES public.ActivityReport.id
+  * userId : integer(32) REFERENCES public.User.id
+  status: enum
```
[the remark this answers is not in the document]
Comment (dcloud) [reply docs/logical_data_model.md:204]: Also, would this model have `createdAt` and `updatedAt` fields to capture when an approver was added and when they updated their status?

**Review by dcloud — COMMENTED** [review]
This approach seems solid. I will go on record as being a fan of a future data migration and model changes since that will help with data uniformity. I should also go on record for being a fan of `calculatedApprovalStatus` (I'm  😄.

It sounds like in the future `activityReport.status` will only have the value of "draft", "submitted", or "deleted" (assuming we migrate past reports).  `activityReportApprover.status` would be either null, `needs_action` or `approved`, which makes sense. 

I think at present someone could select an approver and save a report without submitting it, so there is a small chance there will be draft ARs with a single approver selected. Those theoretical reporters will just reselect approvers when they submit from the new multi-select menu.

File: docs/logical_data_model.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (SarahJaine) [reply docs/logical_data_model.md:204]: Oh yes, definitely. That was an accident. Good catch.

**Comment by SarahJaine** [comment]
> 2\. We migrate all reports over to the multi-approver style (create `ActivityReportApprover` records for all non-legacy reports, approved or not) in a migration at some point before multi-approvers hits the frontend

Sorry, I didn't write out my thinking fully. I'm proposing the "get activity report" API response includes two news parameters: `activityReportApprover` (an array of manager approval objects) and `calculatedApprovalStatus`. The frontend will be updated to expect/use these two new fields from the api response. The handler for the api endpoint will take care of filling in the `activityReportApprover` array and `calculatedApprovalStatus` this for old/single approver reports-- so the api response will always looks the same whether it's old or new data.  Later, if we want to move data around, we can do that, but it will not involve any changes to front end code. 

There's some details on what will go behind a feature flag and what the merge cadence needs to be in my author comment.

**Comment by jasalisbury** [comment]
@SarahJaine @kryswisnaskas can we close this PR?

**Comment by SarahJaine** [comment]
This conversation was fruitful. Final changes are in different PR https://github.com/adhocteam/Head-Start-TTADP/pull/307.

**Pull request closed without merging**